firebase-firestore/src/main/java/com/google/firebase/firestore/model/QueryPlanner.java

```diff
@@ -0,0 +1,317 @@
+    String targetCollection =
+        target.getCollectionGroup() != null
+            ? target.getCollectionGroup()
+            : target.getPath().getLastSegment();
+    hardAssert(index.getCollectionId().equals(targetCollection), "Collection IDs do not match");
+
+    // Queries without filters only use the orderBy() clause for filtering. This is better than
+    // doing a collection-level scan as an orderBy constraint filters out documents that do not
+    // contain the fields being ordered by.
+    if (target.getFilters().isEmpty()) {
+      return processOrderBy(index, new ArrayList<>());
+    }
+
+    // A query that only contains equality clauses can be served by a zig-zag merge join on the
+    // backend. While the SDK can do something similar, for now we pick one of the equalities
+    // and post-filter results in LocalDocumentsView.
+    // TODO(indexing): Determine whether we should implement a zig-zag join
+    if (canUseMergeJoin(target)) {
+      return processMergeJoin(index);
+    }
+
+    return processFilterAndOrderBy(index);
+  }
+
+  /**
+   * Returns the portion of the given `index` that matches both the current target's filters and its
+   * orderBys.
+   */
+  private FieldIndex processFilterAndOrderBy(FieldIndex index) {
+    List<FieldIndex.Segment> expectedSegments = new ArrayList<>();
+
+    // Gather the required index segments for the target's filters. Inequality filters (<=, <, >,
+    // >=, !=, not in) and equality filters (=, IN) can be served by an ORDERED index.
+    // ARRAY_CONTAINS and ARRAY_CONTAINS_ANY require a CONTAINS segment.
+    Set<FieldPath> processedFieldPaths = new HashSet<>();
+    for (Filter filter : target.getFilters()) {
+      // If a field is used in more than one filter (e.g. `where('a', '>', 1).where('a', '<', 3)`)
+      // only add one indexing clause.
+      if (processedFieldPaths.contains(filter.getField())) {
+        continue;
+      }
+      processedFieldPaths.add(filter.getField());
+      hardAssert(filter instanceof FieldFilter, "Only FieldFilters are supported");
+      expectedSegments.add(convertToIndexSegment((FieldFilter) filter));
+    }
+
+    return processOrderBy(index, expectedSegments);
+  }
+
+  /**
+   * Processes a query that can be served by a merge join. These queries can be served by partial
+   * indices that serve any of the equalities that the target is filtering by.
+   */
+  private FieldIndex processMergeJoin(FieldIndex index) {
+    FieldIndex bestMatch = new FieldIndex(index.getCollectionId());
+
+    if (memoizedFilterPermutations == null) {
+      // Build a set of filters that represent all possible combinations (all subsets in all
+      // orders).
+      // For a simple query with two equality clauses this means that we return four different
+      // filter combinations: filter1, filter2, filter1+filter2 and filter2+filter1
+      // Note that while this is potentially expensive, in reality the number of filters should be
```

Comment:
Maybe backend can handle this, for SDK this still make me nervous. I know 8 filters must be rare, may never happen, but 8! == 40320. This will consume a large number of CPU cycles to compute, and memory.

Can we set a hard cap, that we would only compute full sets for the first N (7 likely) filters? This should be enough to get a good performance..and we can sleep better too.